etc/run-clang-tidy-cached.cc

```diff
@@ -0,0 +1,627 @@
+  bool is_bazel_project = false;
+
+  // If the compilation DB or toplevel_build_file changed in timestamp, it
+  // might be worthwhile revisiting sources that previously had issues.
+  // This flag enables that.
+  //
+  // It is good to set once the project is 'clean' and there are only a
+  // few problematic sources to begin with, otherwise every update of the
+  // compilation DB will re-trigger revisiting all of them.
+  bool revisit_brokenfiles_if_build_config_newer = true;
+
+  // Revisit a source file if any of its include files changed content. Say
+  // foo.cc includes bar.h. Reprocess foo.cc with clang-tidy when bar.h changed,
+  // even if foo.cc is unchanged. This will find issues in which foo.cc relies
+  // on something bar.h provides.
+  // Usually good to keep on, but it can result in situations in which a header
+  // that is included by a lot of other files results in lots of reprocessing.
+  bool revisit_if_any_include_changes = true;
+
+  // Clang tidy configuration: clang tidy files with checks. This can be
+  // overriden with environment variable CLANG_TIDY_CONFIG
+  std::string_view clang_tidy_file = ".clang-tidy";
+};
+
+// --------------[ Project-specific configuration ]--------------
+static constexpr ConfigValues kConfig = {
+    // Configuration for your project goes here. Makes it simple to switch to an
+    // updated run-clang-tidy-cached.cc: only need to transfer this block.
+    .start_dir = "src/",
+    .file_exclude_re
+    = "src/sta/"  // Don't check 3rd-party submodule
+      "|codeGenerator/templates/",
+    .revisit_brokenfiles_if_build_config_newer = false,
+};
+// --------------------------------------------------------------
+
+// More clang-tidy config.
+static constexpr std::string_view kExtraArgs[]
+    = {"-Wno-unknown-pragmas", "-Wno-unknown-warning-option"};
+
+// All the extensions we consider
+inline bool IsOneOf(std::string_view s,
+                    std::initializer_list<std::string_view> options)
+{
+  return std::any_of(options.begin(),
+                     options.end(),
+                     [s](std::string_view value) { return value == s; });
+}
+
+// Files that look like relevant include files.
+inline bool IsIncludeExtension(std::string_view extension)
+{
+  return IsOneOf(extension, {".h", ".hpp", ".hxx", ".inl"});
+}
+
+// Filter for source files to be considered.
+inline bool ConsiderExtension(const std::string_view ext)
+{
+  return IsOneOf(ext, {".cc", ".cpp", ".cxx"}) || IsIncludeExtension(ext);
+}
+
+namespace {
+
+namespace fs = std::filesystem;
+using file_time = std::filesystem::file_time_type;
+using ReIt = std::sregex_iterator;
+using hash_t = uint64_t;
+using filepath_contenthash_t = std::pair<fs::path, hash_t>;
+
+// Some helpers
+std::string GetContent(FILE* f)
+{
+  std::string result;
+  if (!f) {
+    return result;  // ¯\_(ツ)_/¯ best effort.
+  }
+  char buf[4096];
+  while (const size_t r = fread(buf, 1, sizeof(buf), f)) {
+    result.append(buf, r);
+  }
```

Comment:
Why old style C io ? GPT gives
```
    std::ostringstream buffer;
    buffer << file.rdbuf();  // Read the file into the stream buffer
    std::string content = buffer.str();  // Convert the buffer to a string
```

Reply:
std::stream stuff is significantly slower than the `fread()` and stuff. that is why I prefer that.